### ISSUE TITLE
Cannot save customer in admin if date of birth is empty (PHP 7.1)

### ISSUE THREAD

**Issue by Yonn-Trimoreau** [magento/magento2#13311]
### Preconditions
1. Magento 2.2.2
2. PHP 7.1.12

### Steps to reproduce
1. Edit a customer in Magento's backoffice
2. Leave the date of birth field empty
3. Save the customer
4. You should see an error message "Something went wrong while saving the customer."

### Expected result
1. Customer should be saved successfully even if date of birth is empty

### Actual result
1. Here is the full stack trace:
```
#0 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/zendframework1/library/Zend/Locale/Math/PhpMath.php(94): Magento\Framework\App\ErrorHandler->handler(2, 'A non-numeric v...', '/home/ytu/Devel...', 94, Array)
#1 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/zendframework1/library/Zend/Locale/Format.php(385): Zend_Locale_Math_PhpMath::Sub('0', '', 0)
#2 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/zendframework1/library/Zend/Locale/Format.php(662): Zend_Locale_Format::toNumber('', Array)
#3 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/zendframework1/library/Zend/Filter/NormalizedToLocalized.php(106): Zend_Locale_Format::toFloat('', Array)
#4 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/Data/Form/Filter/Date.php(65): Zend_Filter_NormalizedToLocalized->filter('')
#5 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/module-customer/Model/Metadata/Form/AbstractData.php(196): Magento\Framework\Data\Form\Filter\Date->inputFilter('')
#6 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/module-customer/Model/Metadata/Form/Date.php(20): Magento\Customer\Model\Metadata\Form\AbstractData->_applyInputFilter('')
#7 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/module-customer/Model/Metadata/Form.php(242): Magento\Customer\Model\Metadata\Form\Date->extractValue(Object(Magento\Framework\App\Request\Http))
#8 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/module-customer/Controller/Adminhtml/Index/Save.php(74): Magento\Customer\Model\Metadata\Form->extractData(Object(Magento\Framework\App\Request\Http), 'customer')
#9 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/module-customer/Controller/Adminhtml/Index/Save.php(44): Magento\Customer\Controller\Adminhtml\Index\Save->_extractData('adminhtml_custo...', 'customer', Array, 'customer')
#10 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/module-customer/Controller/Adminhtml/Index/Save.php(186): Magento\Customer\Controller\Adminhtml\Index\Save->_extractCustomerData()
#11 /home/ytu/Devel/eclipse-workspace/nmk-magento/generated/code/Magento/Customer/Controller/Adminhtml/Index/Save/Interceptor.php(24): Magento\Customer\Controller\Adminhtml\Index\Save->execute()
#12 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/App/Action/Action.php(107): Magento\Customer\Controller\Adminhtml\Index\Save\Interceptor->execute()
#13 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/module-backend/App/AbstractAction.php(229): Magento\Framework\App\Action\Action->dispatch(Object(Magento\Framework\App\Request\Http))
#14 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Backend\App\AbstractAction->dispatch(Object(Magento\Framework\App\Request\Http))
#15 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Customer\Controller\Adminhtml\Index\Save\Interceptor->___callParent('dispatch', Array)
#16 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/module-backend/App/Action/Plugin/Authentication.php(143): Magento\Customer\Controller\Adminhtml\Index\Save\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#17 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/Interception/Interceptor.php(135): Magento\Backend\App\Action\Plugin\Authentication->aroundDispatch(Object(Magento\Customer\Controller\Adminhtml\Index\Save\Interceptor), Object(Closure), Object(Magento\Framework\App\Request\Http))
#18 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Customer\Controller\Adminhtml\Index\Save\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#19 /home/ytu/Devel/eclipse-workspace/nmk-magento/generated/code/Magento/Customer/Controller/Adminhtml/Index/Save/Interceptor.php(39): Magento\Customer\Controller\Adminhtml\Index\Save\Interceptor->___callPlugins('dispatch', Array, NULL)
#20 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/App/FrontController.php(55): Magento\Customer\Controller\Adminhtml\Index\Save\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#21 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\App\FrontController->dispatch(Object(Magento\Framework\App\Request\Http))
#22 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\App\FrontController\Interceptor->___callParent('dispatch', Array)
#23 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\App\FrontController\Interceptor->Magento\Framework\Interception\{closure}(Object(Magento\Framework\App\Request\Http))
#24 /home/ytu/Devel/eclipse-workspace/nmk-magento/generated/code/Magento/Framework/App/FrontController/Interceptor.php(26): Magento\Framework\App\FrontController\Interceptor->___callPlugins('dispatch', Array, Array)
#25 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/App/Http.php(135): Magento\Framework\App\FrontController\Interceptor->dispatch(Object(Magento\Framework\App\Request\Http))
#26 /home/ytu/Devel/eclipse-workspace/nmk-magento/generated/code/Magento/Framework/App/Http/Interceptor.php(24): Magento\Framework\App\Http->launch()
#27 /home/ytu/Devel/eclipse-workspace/nmk-magento/vendor/magento/framework/App/Bootstrap.php(256): Magento\Framework\App\Http\Interceptor->launch()
#28 /home/ytu/Devel/eclipse-workspace/nmk-magento/pub/index.php(43): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#29 {main}
```
2. My investigations led me to an issue with Zend's date filter when you provide an empty string.

3. Maybe you should check if value is an empty string before applying date filter in `Magento_Customer\Model\Metadata\Form\AbstractData` in `_applyInputFilter` function at line `196`

**Comment by magento-engcom-team** [comment]
@Yonn-Trimoreau Thank you for your report. We weren't able to reproduce your issue on the clean instance of Magento. Please clarify what fields where you editing and also let us know if you are using any 3'rd party modules? Thank you

**Comment by Yonn-Trimoreau** [comment]
As you can see in the stack, there's no 3rd party modules involved. I've tested the issue on a clean Magento install.

I wasn't editing any field. I've just opened the customer edit page and saved it as is (no date of birth was entered when I previously registered).

Please, you should retry using **PHP 7.1.12** and with an **empty date of birth**.

**Comment by magento-engcom-team** [comment]
Hi again, @Yonn-Trimoreau. Here is our short video of how we were trying to reproduce Your issue with the php version You have mentioned. We were still not able to reproduce it. Please watch the video and tell us where our steps differ. Thank You.
[Dateofbirth.zip](https://github.com/magento/magento2/files/1699322/Dateofbirth.zip)

**Comment by Yonn-Trimoreau** [comment]
You're doing exactly the same thing as I do :+1: 

Could you try to update your php.ini's `error_reporting` value to `E_ALL & ~E_DEPRECATED & ~E_STRICT` and then retry? (this is what I've set in my php.ini, and it's the default value for php7.1-fpm on debian distros)

Issue seems to be linked to PHP 7.1's new E_WARNING error flag. (more infos [here](https://stackoverflow.com/a/42044413/2110948) and [here](http://php.net/manual/en/migration71.other-changes.php#migration71.other-changes.apprise-on-arithmetic-with-invalid-strings))

Zend should have failed fast when sending an empty string to its date filter, or should have just returned the empty string as is before sending it to its filtering machinery... Still, the issue is happening inside of Magento 2, and the fix is not that hard to implement...

**Comment by TomashKhamlai** [comment]
@Yonn-Trimoreau, what locale was used in Magento?
When you look into Stores > Settings > Configuration > Customers > Customer Configuration > Name and Address Options the value for 'Show Date of Birth' is set to 'No', 'Optional' or 'Required'?

It is not reproducible for me, but I read that lot of people(not only Magento users have the similar problems), you can find the same problem [here](https://github.com/dompdf/dompdf/issues/1272) and [there](http://php.net/manual/en/migration71.other-changes.php).

In one of the article, there was a workaround to use ext-bcmath. I hope that you have a local server to play with and you WON'T do any experiments on your real server.
Would you mind if I ask you to do the following steps:
1) add to your composer.json the line that was added [here](https://github.com/magento/magento2/pull/12768/commits/88684359919e3e77249772d091009a345be1f634)
2) install PHP extension php7.1-bcmath [explained here](https://stackoverflow.com/questions/43905580/php-7-1-bcmath-not-working)
3) switch to developer mode by issuing this line of code:
`php bin/magento deploy:mode:set developer`
4) update your composer dependencies
composer update
5) and perform the upgrade
`php bin/magento setup:upgrade`

I do not know if this will help you. The only thing I know that php7.1 will show the similar error to your one if it meets the line like this: "'1b' + 'something';" while php7.0 will not.

**Comment by Yonn-Trimoreau** [comment]
Hi @TomashKhamlai,

The locale I'm using for this instance is `fr_FR` but I've juste tried to reproduce the issue when locale is set to `en_US`. Issue is still present.

`Show Date of Birth` is set to No on this instance, but when I initially registered it was set to Optional (as Magento's default). I've juste tried to register three new accounts, each of them with a different value set for `Show Date of Birth`. Issue is still present when editing any of them.

My production server is using PHP 7.0 and therefore the issue is not raised on this instance.

I could try your solution that includes installing `php7.1-bcmath` and `ext-bcmath`. But I'm sure we can fix this issue without adding an extra (not justified?) dependency to Magento 2 core...

As I previously wrote in my issue summary: Maybe we should check if value is an empty string before applying date filter in `Magento_Customer\Model\Metadata\Form\AbstractData` in `_applyInputFilter` function at line `196`

If date of birth is an empty string (or any date that needs to go through Zend filtering in Magento), the empty string should be returned as is. Since, in any case, when the date filter is not raising an issue and we send it an empty string, it returns the same empty string. Nothing would be broken, no extra dependency added and issue is fixed once and for all.

**Comment by Yonn-Trimoreau** [comment]
Or maybe more `magento/framework/Data/Form/Filter/Date.php`... Let me make a PR.

**Comment by Yonn-Trimoreau** [comment]
Issue has been fixed!! https://github.com/magento/magento2/commit/799199e149b749890a8ed15b27d751cc7c57dade

**Comment by Yonn-Trimoreau** [comment]
Referenced issue: https://github.com/magento/magento2/issues/12146

**Comment by magento-engcom-team** [comment]
@Yonn-Trimoreau, thank you for your report.
This seems to be correct Magento behavior. Please refer to the [Community Forums](https://community.magento.com/) or the [Magento Stack Exchange](https://magento.stackexchange.com/) site for advice or general discussion about this.
Otherwise you may submit Pull Request with the suggested changes.

**Comment by barryvdh** [comment]
But it's not fixed in the current version, is it? The commit you reference is from before the 2.2.2 release, or was it not included?

**Comment by Yonn-Trimoreau** [comment]
I can't see any version tag for this commit. Maybe it will be integrated in 2.2.3

You can still temporarily edit vendor code to fix it in 2.2.2.. Or override & extend the file if you're already in production.

**Comment by lano-vargas** [comment]
@magento-engcom-team  Same issue here. 2.2.2

**main.CRITICAL: Exception message: Warning: A non-numeric value encountered in .../vendor/magento/zendframework1/library/Zend/Locale/Math/PhpMath.php on line 94**

Once I added DOB I was able so save/create new customer from backend.
